### DISCUSSION TITLE
using SPARQL Anything as a java library

### DISCUSSION THREAD

**Discussion by justin2004** [SPARQL-Anything/sparql.anything#288] [Q&A]
I have a Clojure/Java project where I use Apache Jena to run SPARQL constructs.
I'd like to be able to use service clauses with the `x-sparql-anything:` scheme. 

If I don't put the SPARQL Anything jar on the class path and I use Jena to execute this:

```sparql
PREFIX  fx:   <http://sparql.xyz/facade-x/ns/>
                                    construct {?s ?p ?o}
                                    where {
                                    service <x-sparql-anything:> {
                                    fx:properties fx:location "/tmp/la.json" .
                                    ?s ?p ?o
                                    }}
```

I get this error (as expected):
```
Execution error (IllegalArgumentException) at jdk.internal.net.http.common.Utils/newIAE (Utils.java:280).
invalid URI scheme x-sparql-anything
```

Next, if I put the SPARQL Anything jar on the class path and evaluate this:

```clojure
(org.apache.jena.sparql.engine.main.QC/setFactory (org.apache.jena.query.ARQ/getContext)
                                                  com.github.sparqlanything.engine.FacadeX/ExecutorFactory)
```

which is equivalent to:

```java
QC.setFactory(ARQ.getContext(), FacadeX.ExecutorFactory);
```

Then I execute that construct again I get:

```
Execution error (QueryExceptionHTTP) at org.apache.jena.sparql.engine.http.HttpQuery/execGet (HttpQuery.java:324).
Malformed URL: java.net.MalformedURLException: unknown protocol: x-sparql-anything
```

I'd like to know how to tell Apache Jena to use FacadeXOpExecutor when that protocol is encountered. Any tips on how I might do that?

My current guess as to why this isn't working... I think it is related to this... the static initializer in the FacadeX class (which registers functions and magic properties, etc.) might not be getting executed?

**Comment by luigi-asprino** [comment]
Hi,

> I'd like to know how to tell Apache Jena to use FacadeXOpExecutor when that protocol is encountered. Any tips on how I might do that?

It should be this

```
QC.setFactory(ARQ.getContext(), FacadeX.ExecutorFactory);
```

The error doesn't seem related to the static initialiser (static block), otherwise, you should get something related to the fact that the engine is not able to find a triplifier for the source. 
IMO, the problem is that the operation is not delegated to the FXOpExecutor (maybe because FacadeX.ExecutorFactory is not initialised) 
You can try instantiating the OpExecutorFactory explicitly

```
OpExecutorFactory exFactory = new OpExecutorFactory() {
		@Override
		public OpExecutor create(ExecutionContext execCxt) {
			return new FacadeXOpExecutor(execCxt);
		}
	};

QC.setFactory(ARQ.getContext(), exFactory);

```

**Reply by justin2004** [reply]
thanks for the tip, @luigi-asprino .

i tried your suggestion and i got the same error. also, i noticed this...

when the factory is nil
```clojure
(org.apache.jena.sparql.engine.main.QC/getFactory (org.apache.jena.query.ARQ/getContext)) ; => nil
```
i get
```
Execution error (QueryExceptionHTTP) at org.apache.jena.sparql.engine.http.HttpQuery/execGet (HttpQuery.java:324).
Malformed URL: java.net.MalformedURLException: unknown protocol: x-sparql-anything
```

and when the factory is not nil 
```clojure
(org.apache.jena.sparql.engine.main.QC/getFactory (org.apache.jena.query.ARQ/getContext))
; => #object[com.github.sparqlanything.engine.FacadeX$1 0x5fa0d03 com.github.sparqlanything.engine.FacadeX$1@5fa0d03]
```
i get the same error as above.

i think i might put a break point [here](https://github.com/SPARQL-Anything/sparql.anything/blob/v0.8-DEV/sparql-anything-engine/src/main/java/com/github/sparqlanything/engine/FacadeXOpExecutor.java#L499) (or somewhere in that file) to see if execution even passes through there when i run a construct.
